# Coding Exercises (Part 1)

## Exercise 3: Integers, Floats, Lists and built-in functions

1. __Create and save objects__ of your choice with the following __data types__. Check the data type with the appropriate function!

integer

In [1]:
a = 10

In [2]:
type(a)

int

float

In [3]:
b = 9.18

In [4]:
type(b)

float

string

In [5]:
c = "pranav"

In [6]:
type(c)

str

bool

In [7]:
d = True

In [8]:
type(d)

bool

2. Determine the __sum__, the __number of elements__ and __maximum value__ in list l. Use __built-in functions__!

In [9]:
# run the cell
l = [-300, 40, 140, -430, 85, 280, 0, -560, 340]

In [10]:
#sum
sum(l)

-405

(result: -405)

In [11]:
#number of elements
len(l)

9

(result: 9)

In [12]:
#maximum
max(l)

340

(result: 340)

3. Determine the __number of elements__ and __the alphabetically first value__ in string s. Use __built-in functions__!

In [13]:
#run the cell
s = "myrmecophilous"

In [14]:
#number of elements
len(s)

14

(result: 14)

In [15]:
#alphabetically first value
min(s)

'c'

(result: "c")

4. __Check__ whether the following equation is __True__ in Python: 0.2 + 0.2 + 0.2 = 0.6

In [16]:
0.2 + 0.2 + 0.2 == 0.6

False

5. Get the __first 20 decimals__ of 0.2 and 0.6 and explain the result of exercise 4.

In [17]:
format(0.2, ".20f")

'0.20000000000000001110'

(result: '0.20000000000000001110')

In [18]:
format(0.6, ".20f")

'0.59999999999999997780'

(result: '0.59999999999999997780')

6. __Round__ 1.56456 to __2 decimals__! Check the __first 20 decimals__ of the rounded float. Does rounding guarantee exact numbers in Python?

In [19]:
round(1.56456, 2)

1.56

(result: 1.56)

In [20]:
format(1.56, ".20f")

'1.56000000000000005329'

(result: '1.56000000000000005329')

7. __Round__ 2,320,576 to the next __multiple of 1,000__!

In [21]:
round(2320576, -3)

2321000

(result: 2321000)

We have the list cf with cashflows starting today until year 6:

In [22]:
# run the cell
cf = [-2000, 200, 300, 500, 750, 1000, 1500]
cf

[-2000, 200, 300, 500, 750, 1000, 1500]

8. Create the following __slices__ of cf:

first two elements

In [23]:
cf[:2]

[-2000, 200]

the elements 300, 500 and 750

In [24]:
cf[2:5]

[300, 500, 750]

the last three elements

In [25]:
cf[-3:]

[750, 1000, 1500]

all elements, except the last element

In [26]:
cf[:-1]

[-2000, 200, 300, 500, 750, 1000]

every second element (starting from first element)

In [27]:
cf[::2]

[-2000, 300, 750, 1500]

9. __Add__ the cashflow 1200 for year 7 to list cf. __Change__ cf in memory!

In [28]:
cf.append(1200)


In [29]:
cf

[-2000, 200, 300, 500, 750, 1000, 1500, 1200]

10. __Change__ the cashflow in t2 (300) to 400. __Change__ cf in memory!

In [30]:
cf[2] = 400

In [31]:
cf

[-2000, 200, 400, 500, 750, 1000, 1500, 1200]

11. __Remove__ the last three years of cf (with slicing and variable reassignment )

In [32]:
cf = cf[:-3]

In [33]:
cf

[-2000, 200, 400, 500, 750]

12. __Add 20__ to each and every element of cf. Create the new list __cf_new__!

In [34]:
cf_new = []
for i in cf:
    cf_new.append(i + 20)
cf_new

[-1980, 220, 420, 520, 770]

(result: [-1980, 220, 420, 520, 770])

13. __Sort__ cf_new from __high to low__! __Change__ cf_new in memory!

In [35]:
cf_new.sort(reverse = True)

In [36]:
cf_new

[770, 520, 420, 220, -1980]

(result: [770, 520, 420, 220, -1980])

Initially, we have two projects with identical cashflows:

In [37]:
#run the cell
cf1 = [-50, 10, 30, 50]
cf2 = cf1

Then, the first cashflow of project 1 changes to -60.

In [38]:
#run the cell
cf1[0] = -60
cf1

[-60, 10, 30, 50]

14. Check, whether __cf2 changed__ as well!

In [39]:
cf2

[-60, 10, 30, 50]

15. __Rewrite__ the above code and make sure that __cf2 does not change__! Follow __best practices__!

In [40]:
cf1 = [-50, 10, 30, 50]
cf2 = cf1.copy()

In [41]:
cf1[0] = -60
cf1

[-60, 10, 30, 50]

In [42]:
cf2

[-50, 10, 30, 50]

# Well Done!

---------------------------------------------------------------------------------------------------------------------